store/tikv/mock-tikv/mvcc_leveldb.go

```diff
@@ -0,0 +1,724 @@
+	startKey := mvccEncode(key, lockVer)
+	iter := newIterator(mvcc.db, &util.Range{
+		Start: startKey,
+	})
+	defer iter.Release()
+
+	return getValue(iter, key, startTS, isoLevel)
+}
+
+func getValue(iter *Iterator, key []byte, startTS uint64, isoLevel kvrpcpb.IsolationLevel) ([]byte, error) {
+	dec1 := lockDecoder{expectKey: key}
+	ok, err := dec1.Decode(iter)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if ok {
+		if isoLevel == kvrpcpb.IsolationLevel_SI && dec1.lock.startTS <= startTS {
+			return nil, dec1.lock.lockErr(key)
+		}
+	}
+
+	dec2 := valueDecoder{expectKey: key}
+	for iter.Valid() {
+		ok, err := dec2.Decode(iter)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if !ok {
+			break
+		}
+
+		value := &dec2.value
+		if value.valueType == typeRollback {
+			continue
+		}
+		if value.commitTS <= startTS {
+			if value.valueType == typeDelete {
+				return nil, nil
+			}
+			return value.value, nil
+		}
+	}
+	return nil, nil
+}
+
+// BatchGet implements the MVCCStore interface.
+func (mvcc *MVCCLevelDB) BatchGet(ks [][]byte, startTS uint64, isoLevel kvrpcpb.IsolationLevel) []Pair {
+	mvcc.mu.RLock()
+	defer mvcc.mu.RUnlock()
+
+	var pairs []Pair
+	for _, k := range ks {
+		v, err := mvcc.getValue(k, startTS, isoLevel)
+		if v == nil && err == nil {
+			continue
+		}
+		pairs = append(pairs, Pair{
+			Key:   k,
+			Value: v,
+			Err:   err,
+		})
+	}
+	return pairs
+}
+
+// Scan implements the MVCCStore interface.
+func (mvcc *MVCCLevelDB) Scan(startKey, endKey []byte, limit int, startTS uint64, isoLevel kvrpcpb.IsolationLevel) []Pair {
+	mvcc.mu.RLock()
+	defer mvcc.mu.RUnlock()
+
+	iter, currKey, err := newScanIterator(mvcc.db, startKey, endKey)
+	defer iter.Release()
+	if err != nil {
+		log.Error("scan new iterator fail:", errors.ErrorStack(err))
+		return nil
+	}
+
+	ok := true
+	var pairs []Pair
+	for len(pairs) < limit && ok {
+		value, err := getValue(iter, currKey, startTS, isoLevel)
+		if err != nil {
+			pairs = append(pairs, Pair{
+				Key: currKey,
+				Err: errors.Trace(err),
+			})
+		}
+		if value != nil {
+			pairs = append(pairs, Pair{
+				Key:   currKey,
+				Value: value,
+			})
+		}
+
+		skip := skipDecoder{currKey}
+		ok, err = skip.Decode(iter)
+		if err != nil {
+			log.Error("seek to next key error:", errors.ErrorStack(err))
+			break
+		}
+		currKey = skip.currKey
+	}
+	return pairs
+}
+
+// ReverseScan implements the MVCCStore interface.
+func (mvcc *MVCCLevelDB) ReverseScan(startKey, endKey []byte, limit int, startTS uint64, isoLevel kvrpcpb.IsolationLevel) []Pair {
+	return nil
+}
+
+// Prewrite implements the MVCCStore interface.
+func (mvcc *MVCCLevelDB) Prewrite(mutations []*kvrpcpb.Mutation, primary []byte, startTS uint64, ttl uint64) []error {
+	mvcc.mu.Lock()
+	defer mvcc.mu.Unlock()
+
+	anyError := false
+	errs := make([]error, 0, len(mutations))
+	for _, m := range mutations {
+		err := prewriteMutation(mvcc.db, m, startTS, primary, ttl)
```

Comment:
If some mutation returns error, others will be written to DB.
tikv's behavior to generate a writeBatch, and the writeBatch is written to DB only if no error happens.